### PR TITLE
Remove showUI to reduce friction fixes issue:2361

### PR DESCRIPTION
**Pull request by shahbaz17**
This PR fixes https://github.com/redwoodjs/redwood/issues/2361

Removed `showUI` to reduce friction/confusion as to what has happened after clicking login/sign up.

### CODE REVIEW

**Comment by jtoar** [comment]
Thanks for taking this one on @shahbaz17!

File: packages/auth/src/authClients/magicLink.ts
Comment (jtoar) [inline packages/auth/src/authClients/magicLink.ts:8]: One thing @dthyresson said in the issue this PR addresses, #2361:

> That way, only if the developer doesn't want the confirm message, then they can explicitly set showUI: false

If we remove `showUI` entirely, the developer won't be able to opt out right?

Comment (shahbaz17) [reply packages/auth/src/authClients/magicLink.ts:8]: Yes, that's true.

it removes the `showUI` entirely.

That is why I made a PR on https://github.com/redwoodjs/playground-auth/pull/60 which would have worked with the current implementation of Magic in redwood.js.

It is about addressing the parameters, [loginWithMagicLink](https://docs.magic.link/client-sdk/web/api-reference#loginwithmagiclink) addresses `email`, `showUI`, but not the `redirectURI`. Similarly the [getIdToken](https://docs.magic.link/client-sdk/web/api-reference#getidtoken) doesn’t address the `lifespan` parameter.

https://github.com/redwoodjs/redwood/blob/main/packages/auth/src/authClients/magicLink.ts#L18

https://github.com/redwoodjs/redwood/blob/main/packages/auth/src/authClients/magicLink.ts#L27

Comment (dthyresson) [reply packages/auth/src/authClients/magicLink.ts:8]: Sorry, I am confused. Shouldn't we default showUI to false?

If the point is to "reduce friction" I assume you do not want to have the popup message? Thus it's false by default and a developer can still opt-in to true to show message?

```js
export const magicLink = (client: MagicLink): AuthClientMagicLink => {
  return {
    type: 'magicLink',
    client,
    login: async ({ email, showUI = false }) =>
      await client.auth.loginWithMagicLink({ email, showUI }),
    logout: async () => {
      await client.user.logout()
    },
    signup: async ({ email, showUI = false }) =>
      await client.auth.loginWithMagicLink({ email, showUI }),
    getToken: async () => await client.user.getIdToken(),
    getUserMetadata: async () =>
      (await client.user.isLoggedIn()) ? await client.user.getMetadata() : null,
  }
}
```

**Comment by thedavidprice** [comment]
Hi all! Checking on the status here. Seems like the conversation momentum got lost. Worth picking back up?

**Comment by shahbaz17** [comment]
Hey @thedavidprice 

Thanks for checking in.

I believe this PR is waiting for review.

**Comment by thedavidprice** [comment]
Thanks @shahbaz17 In that case, I want to make sure you saw this last review comment from @dthyresson:
https://github.com/redwoodjs/redwood/pull/2629#discussion_r643190647

Appreciate the work here!

File: packages/auth/src/authClients/magicLink.ts
Comment (doesnotexist) [reply packages/auth/src/authClients/magicLink.ts:8]: Hrmm, well I would have interpreted "reduce friction" here to mean something more like "batteries included" aka reduce developer time to increase adoption.   If showUI defaults to false, developers then need to bring their own UI to display to users so they know that something is happening. Whereas with showUI true, the magic sdk provides that for you with the iframe overlay shown above, which also hides itself once login is completed.

**Review by thedavidprice — CHANGES_REQUESTED** [review]
Just to be clear I understand the scop/context/intention of this PR, I'm copy+pasting from the associated Issue and Playground-auth PR:

From https://github.com/redwoodjs/playground-auth/pull/60
> I do recommend showing the pop-up UI, it gives users useful information on what to do.
> 
> What I suggest is not adding `showUI`:`true` to `await logIn({ email, showUI: true })`. 
> 
> `await client.auth.loginWithMagicLink({ email, showUI })` when written as `await client.auth.loginWithMagicLink({ email })` still shows the pop-up UI, but If you don’t want that pop-up to be shown, you can pass `showUI`: `false`.
> 
> `showUI` is `true` by default.

...which is restated here https://github.com/redwoodjs/redwood/issues/2361

> has suggested that the default behavior of Magiclink should always be to show the UI as in:
> 
> image
> 
> That way, only if the developer doesn't want the confirm message, then they can explicitly set showUI: false

---

**Goals of this PR:**
- enforce a default to SHOW the popup
- reduce any confusion a dev might have about 1) best practices and 2) options required to use Magic.link

**My 2 Cents:**
This PR goes too far in completely removing the (optional) option. Understanding that it's both 1) optional and 2) defaults to `true`, I don't think the correct approach is to remove it.

I'd prefer an approach that adds clarity about available options along with the simplest implementation of the recommended defaults. E.g.:
```
    login: async ({ email, showUI = true }) =>
       await client.auth.loginWithMagicLink({ email, showUI}),
```

As a developer reading this code:
- I know the option exists
- I no longer have to look up the default value of showUI
- I implicitly understand the recommendation is to have the UI display
- and I know I can just use `login({email})` in my code

Am I right? What say you all?

**Comment by dthyresson** [comment]
Hi @shahbaz17 this PR has been lingering for awhile, but I was wondering if the entire Magic auth provider should be updated per https://magic.link/docs/admin-sdk/node/api-reference

This way there is a proper decoder that uses

```ts
const { Magic } = require('@magic-sdk/admin');

const mAdmin = new Magic('SECRET_API_KEY');

mAdmin.token.validate;
```

To ensure the token is valid server-side?

Is this something Magic would be able to submit as a PR?

Thanks!

**Comment by shahbaz17** [comment]
I second @thedavidprice point raised [here](https://github.com/redwoodjs/redwood/pull/2629#pullrequestreview-755664087).

Let's use this code in the playground repo and close this PR as it's clear here.

```
    login: async ({ email, showUI = true }) =>
       await client.auth.loginWithMagicLink({ email, showUI}),
```

**Comment by shahbaz17** [comment]
> Hi @shahbaz17 this PR has been lingering for awhile, but I was wondering if the entire Magic auth provider should be updated per https://magic.link/docs/admin-sdk/node/api-reference
> 
> This way there is a proper decoder that uses
> 
> ```ts
> const { Magic } = require('@magic-sdk/admin');
> 
> const mAdmin = new Magic('SECRET_API_KEY');
> 
> mAdmin.token.validate;
> ```
> 
> To ensure the token is valid server-side?
> 
> Is this something Magic would be able to submit as a PR?
> 
> Thanks!

Not sure what you mean here @dthyresson. Are you referring we could use something like this on the client-side as well?

**Comment by shahbaz17** [comment]
@thedavidprice 

I have updated the PR to include your suggestion.

**Comment by doesnotexist** [comment]
@shahbaz17  I think decoder example is referring to api-side because there does not seem to be a token decoder for magic's auth under redwood/packages/api/src/auth/decoders

**Comment by shahbaz17** [comment]
Thanks for clarifying @doesnotexist 

Yes, we can add that under `redwood/packages/api/src/auth/decoders`.

Should I make a separate PR for that?

**Review by dthyresson — COMMENTED** [review]
> Should I make a separate PR for that?

@shahbaz17 Yes, please do. I'll create an issue and assign it to you. Thanks much!

**Comment by shahbaz17** [comment]
Hey @dthyresson 

We are keeping the `showUI?:boolean`. See [this](https://github.com/redwoodjs/redwood/pull/2629/files#diff-3b27c9a3b73454d93e9ae4e7335a43fd5e678869f68862c8647e8f07788dd94fR8).

> Questions:
> 1. Should the UI pop ups always appear. That is the develop should never be able to hide them?

Yes, as the default is true.
```
login: async ({ email, showUI = true }) =>
       await client.auth.loginWithMagicLink({ email, showUI}),
```

And a developer would use `login({email})` to login with UI pop-up, or `login({email, false})` to not show the UI pop-up.

> 2. Does the redirectURI need to be added as an option for login?
`redirectURI` is an optional parameter that Magic will point to after the user clicks their email link. But this will involve working with `loginWithCredential`. see [this](https://magic.link/docs/client-sdk/web/api-reference#loginwithcredential)

`Don't forget to call loginWithCredential at the specified redirect location!`

> 3. Is lifespan missing when retrieving a token?
`loginWithMagicLink` returns a promise. The promise resolves upon authentication request success and rejects with a specific error code if the request fails. The resolved value is a Decentralized ID token with a default 15-minute lifespan.

To get a DID Token of a longer life span, one should use `getIdToken` or `generateIdToken` and pass lifespan as ar. See [this](https://magic.link/docs/client-sdk/web/api-reference#getidtoken)

File: packages/auth/src/authClients/magicLink.ts
Comment (thedavidprice) [reply packages/auth/src/authClients/magicLink.ts]: Code has been updated per my review comments. See https://github.com/redwoodjs/redwood/pull/2629/files

Resolving this conversation

**Review by thedavidprice — APPROVED** [review]

**Comment by thedavidprice** [comment]
@dthyresson my requested changes have been made and look good. I've approved this PR. Once approved by you, please merge.

Thanks @shahbaz17! Grateful for your help with all this.

**Pull request merged**
